src/components/SidePanels/TabPanels/utils.ts

```diff
@@ -189,4 +197,71 @@ export const restoreComponentsToState = (panelData: Record<string, PanelBBox>) =
 
 export const savePanels = (panelData: Record<string, PanelBBox>) => {
   window.localStorage.setItem('panelState', JSON.stringify(panelData));
+};
+
+
+export const reCalcSnappedHeights = (
+  state: Record<string, PanelBBox>,
+  totalHeight: number,
+) => {
+  const newState = { ...state };
+  const leftKeys = Object.keys(newState).filter((key) => !newState[key].detached && newState[key].alignment === Side.left);
+  const rightKeys = Object.keys(newState).filter((key) => !newState[key].detached && newState[key].alignment === Side.right);
+  
+  [leftKeys, rightKeys].forEach(list => {
+    const panelHeight = totalHeight / list.length;
+  
+    list.forEach(panelKey => {
+      newState[panelKey].height = panelHeight;
+    });
+  });
+
+  return newState ;
+};
+
+export const joinPanelColumns = (
+  state: Record<string, PanelBBox>,
+  sameSidePanelKeys: string[],
+  panelAddKey: string,
+  alignment: string,
+  totalHeight: number,
+  width: number,
+) => {
+  const newState = { ...state };
+  let top = 0;
+
+  sameSidePanelKeys.push(panelAddKey);
```

Comment:
```suggestion
export const joinPanelColumns = (
  state: Record<string, PanelBBox>,
  _sameSidePanelKeys: string[],
  panelAddKey: string,
  alignment: string,
  totalHeight: number,
  width: number,
) => {
  const newState = { ...state };
  let top = 0;

  const sameSidePanelKeys = Array.from(new Set([..._sameSidePanelKeys, panelAddKey]));
```
I would do this immutably, otherwise there is a hidden side effect to this function that can produce odd bugs.